File: linked-list/mid_difficulty/2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

   int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        if(head == nullptr){
            return nullptr;
        }

        ListNode* crr = head;
        ListNode* next = head->next;

        while(crr != nullptr && next != nullptr){
            int value = gcd(crr->val, crr->next->val);
            ListNode* newNode = new ListNode(value,next);
            crr->next = newNode;
            crr = newNode->next;
            next = next->next;
        }

        return head;
    }
};

File: Linked-List/2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int GCD(int a, int b){
        while(b != 0){
            int temp = b;
            b = a % b;
            a = temp;
        }

        return a;
    }

    public ListNode insertGreatestCommonDivisors(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }

        ListNode crr = head;

        while(crr != null && crr.next != null){
            int newNodeVal = GCD(crr.val, crr.next.val);
            crr.next = new ListNode(newNodeVal, crr.next);
            crr = crr.next.next;
        }

        return head;
    }
}